### ISSUE TITLE
Handling of concept labels in test data

### ISSUE THREAD

**Issue by ghost** [sifei/2nd-Social-Media-Mining-for-Health-Applications-Shared-Task-at-AMIA-2017#1]
Hi sifei, 
           I have a small doubt regarding **"SMM4H 2017 Medical Concept Normalization Task"**. There are around 470 unique concepts in training set and around 250 unique concepts in test set. There are around 30 concepts which are only in test set but  not in training set.  How did you handle this one?

Waiting for your reply... Thanks in advance.

**Comment by tttr222** [comment]
Hi kalyanks0611,

I'll try to answer your question as best I can. The code for ADR concept normalization task is located here:
https://github.com/tttr222/uknlp_adr_mention_norm

Essentially, we submitted two models, one with external and one without external data. In the case of non-external data, the predicted concepts would be limited to the ones in the training set; hence, test examples with concepts not seen in the training will contribute to overall testing error. Of course, using external training data partially overcomes this issue by expanding the number of concepts that are seen by the model during training.  However, the model we submitted that does not use external data performed best; hence, the code provided is for the version without external data.

Thanks,
Tung

**Comment by ghost** [comment]
Thanks for  the clear explanation.

In this paper "Normalising medical concepts in social media texts by learning semantic representation" ( https://github.com/nutli/concept_normalisation ) , **the author  did like this while encoding the labels into one hot vectors**

categorical_processor = learn.preprocessing.VocabularyProcessor(1)
all_label = y_train_text+y_validation_text+y_test_text
num_labels = len(set(all_label))

y = np.array(list(categorical_processor.fit_transform(all_label)), dtype=np.int32)
y_train = y[0:len(y_train_text)]
y_validation = y[len(y_train_text):len(y_train_text)+len(y_validation_text)]
y_test = y[len(y_train_text)+len(y_validation_text):] 

**In SMM4H MCN task also , can I do like this to overcome the issue of unseen labels in test data ?**

**Comment by tttr222** [comment]
I believe that even if you include test and validation labels in the set of predictable labels, you will get the same results more or less. In addition to training on features, classification models also tend to capture the distribution of the output label space. My thought is that if a particular concept is not seen during training, the model will likely assume it occurs with a 0% probability on unseen data, and therefore not predict it during testing. I think for supervised models, the only way is to introduce external data; ideally, you want a minimum of one example per unique label.

**Comment by ghost** [comment]
Thanks for clear explanation.
